File: python/relay/openai_agents.py
"""Live Relay + OpenAI Agents SDK tool input/output guardrails."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from common import SOURCE_APP, harness, session_id  # noqa: E402


class _Ctx:
    def __init__(self, tool_name: str, arguments: dict) -> None:
        self.tool_name = tool_name
        self.tool_arguments = arguments


class _Data:
    def __init__(self, tool_name: str, arguments: dict, output: object = None) -> None:
        self.context = _Ctx(tool_name, arguments)
        self.output = output


def _is_blocked(result: Any) -> bool:
    if isinstance(result, dict):
        return bool(result.get("tripwire_triggered"))
    if hasattr(result, "tripwire_triggered"):
        return bool(result.tripwire_triggered)
    name = type(result).__name__.lower()
    if "reject" in name:
        return True
    if "allow" in name:
        return False
    return False


def main() -> None:
    h = harness("relay-openai-agents-sample")
    sid = session_id("relay-openai-agents")
    relay = h.relay(surface="openai_agents", source_app=SOURCE_APP)
    on_input = relay.openai_tool_input(session_id=sid)
    on_output = relay.openai_tool_output(session_id=sid)

    allow = on_input(_Data("get_weather", {"city": "SF"}))
    deny = on_input(_Data("shell", {"command": "rm -rf /tmp/x"}))
    out = on_output(_Data("get_weather", {"city": "SF"}, output="Sunny"))

    print(
        json.dumps(
            {
                "example": "relay_openai_agents",
                "sessionId": sid,
                "allowBlocked": _is_blocked(allow),
                "denyBlocked": _is_blocked(deny),
                "outputType": type(out).__name__,
                "wire": (
                    "Agent(..., tool_input_guardrails=[relay.openai_tool_input(...)], "
                    "tool_output_guardrails=[relay.openai_tool_output(...)])"
                ),
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()

File: typescript/relay/openai-agents.ts
/**
 * Relay + OpenAI Agents SDK (`surface: "openai_agents"`).
 *
 * Peers: npm i @openai/agents openai
 * Env:   EXEMPLAR_API_KEY, OPENAI_API_KEY (for a live agent run)
 *
 *   npm run build && npx tsx examples/relay/openai-agents.ts
 *
 * Without OPENAI_API_KEY this still exercises decide/observe via the guardrails
 * against a mocked tool call shape.
 */
import { harness, newSessionId, banner, SOURCE_APP } from "../_helpers.ts";

async function main(): Promise<void> {
  const h = harness("openai-agents-relay-bot");
  const sessionId = newSessionId("relay-openai-agents");
  banner("Relay + OpenAI Agents", sessionId);
  const relay = h.relay({
    surface: "openai_agents",
    sourceApp: SOURCE_APP,
  });

  const inputGuard = relay.openaiToolInput(sessionId);
  const outputGuard = relay.openaiToolOutput(sessionId);

  // Dry-run the guardrail against a tool-call shaped payload (no Agents peer required).
  const dry = await inputGuard({
    tool_name: "shell",
    tool_arguments: { command: "echo hello" },
  });
  console.log("openaiToolInput (dry):", dry);

  await outputGuard({
    tool_name: "shell",
    tool_arguments: { command: "echo hello" },
    output: "hello",
  });
  console.log("openaiToolOutput (dry): observed");

  if (!process.env.OPENAI_API_KEY?.trim()) {
    console.log(
      "\nSkipping live Agent run (set OPENAI_API_KEY + install @openai/agents to enable).",
    );
    console.log("Wire-up sketch:");
    console.log(`
  import { Agent, run, tool } from "@openai/agents";

  const lookup = tool({
    name: "lookup_order",
    description: "Look up an order by id",
    parameters: { type: "object", properties: { orderId: { type: "string" } }, required: ["orderId"] },
    execute: async ({ orderId }) => ({ orderId, status: "shipped" }),
  });

  const agent = new Agent({
    name: "Support",
    instructions: "Use tools when needed.",
    tools: [lookup],
    // Attach Relay decide/observe:
    // toolInputGuardrails: [inputGuard],
    // toolOutputGuardrails: [outputGuard],
  });

  await run(agent, "What is the status of order ORD-1?");
`);
    return;
  }

  try {
    const agents = await import("@openai/agents");
    const lookup = agents.tool({
      name: "lookup_order",
      description: "Look up an order by id",
      parameters: {
        type: "object",
        properties: { orderId: { type: "string" } },
        required: ["orderId"],
        additionalProperties: false,
      },
      strict: true,
      execute: async (input: { orderId: string }) => ({
        orderId: input.orderId,
        status: "shipped",
      }),
    });

    const agent = new agents.Agent({
      name: "Support",
      instructions: "Use lookup_order when the user asks about an order.",
      tools: [lookup],
    });

    // Prefer SDK-native guardrail helpers when present; otherwise rely on decide via dry path above.
    console.log("Running agent (Relay decide still available via openaiToolInput/Output)…");
    const result = await agents.run(agent, "Status of order ORD-42?");
    console.log("agent output:", String((result as { finalOutput?: unknown }).finalOutput ?? result));
  } catch (err) {
    console.warn("Live Agents run skipped:", err instanceof Error ? err.message : err);
    console.warn("Install peers: npm i @openai/agents openai");
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
